### ISSUE TITLE
To use the RSL or not to use the RSL

### ISSUE THREAD

**Issue by adityasaky** [gittuf/gittuf#668]
### Add a description

In many flows, we identify the tip of a reference from its latest entry in the RSL. For example, when creating an approval attestation for a change from `feature` being merged into `main`, we compute the expected merge tree using the commits identified in the latest RSL entries for both branches. @wlynch pointed out that this may be incorrect in cases where the RSL has not been updated, perhaps locally for the approver.

Using the RSL mitigates reference state attacks against the refs in question, so I'm inclined to keep using the RSL and try and maximize RSL entry creation across the board so the latest entry matches the ref as expected. But, do we want to provide a flag? Is there a flow where an approver wants to approve a change from commit Y on the feature branch even though the latest RSL entry is for commit X? Doesn't this indicate that the feature branch has moved when it shouldn't have?

### Relevant log output if the discussion pertains to existing gittuf functionality

```Shell

```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by JustinCappos** [comment]
I need to understand this case better, but in general, my feeling is the
RSL should always represent an append-only log of up-to-date actions.
 This is important to stop the repository, etc. from doing all the various
branch attacks.   Unless there is some critical usability issue this
raises, I'm in favor of not changing the invariant that we leverage the RSL
as much as possible.

**Comment by adityasaky** [comment]
No, I think we want to treat `gittuf/gittuf` and `adityasaky/gittuf` as distinct repositories as far as push activity recorded in the RSL is concerned. They may have distinct policies and rules depending on the context. For example, there may be a long lived fork that serves as a staging area for pushes with their own PRs (+ approvals) that are then upstreamed all at once. This is reminiscent of the kernel's development workflow, where a subsystem maintainer's repo is where changes land before getting pulled into torvalds/linux.

**Comment by adityasaky** [comment]
@JustinCappos, @patzielinski, and I had a quick chat yesterday. For the GitHub app case specifically (and perhaps others like it where we cross the repository boundary), we will not use the RSL to determine the tip of the feature branch. We can likely resort to looking at the present tip of the feature branch only when the repository for the feature branch differs from the repository for the base branch.

I'll leave the issue open for the time being in case we have more thoughts on the general question of when to use the RSL.

**Comment by adityasaky** [comment]
> For the GitHub app case specifically (and perhaps others like it where we cross the repository boundary), we will not use the RSL to determine the tip of the feature branch. We can likely resort to looking at the present tip of the feature branch only when the repository for the feature branch differs from the repository for the base branch.

I've opened #766 for this.
